### ISSUE TITLE
Way to call callback when diff rendering finished

### ISSUE THREAD

**Issue by marcinkuzminski** [wickedest/Mergely#32]
I'm looking for a way to scroll to first change once editor is loaded, and both lhs and rhs sides loaded as well. The problem is that i think there's no callback once the diff calculation is done. 

In case of large files generating diffs takes some time, and there's no nice way to call the scrollToDiff function then.

Am i missing something ?

**Comment by marcinkuzminski** [comment]
Been playing arround the code a little bit, it seems when this might be problematic since calling
.mergely('scrollToDiff', 'prev'); calls _diff function itself, so a callback could create an loop.

Any other idea how to add scroll to first change once the diffs are loaded ?

**Comment by wickedest** [comment]
Hi Marcin,

There is no callback.  The Mergely editor does a similar thing with scrollTo
<https://github.com/wickedest/Mergely/blob/master/editor/editor.js#L193>.
It might work for scrollToDiff.  Generally speaking, wait for the lhs/rhs
editors to be initialized, then call scrollToDiff.  There might be some
contention between the autoupdate <http://www.mergely.com/doc#autoupdate>
(auto-diff calculation) and the scrollToDiff, so you may need to disable it
if initialization is flaky.

Good luck!
Jamie

**Comment by marcinkuzminski** [comment]
Hi Jamie,

Thanks for the the info. The problem is i cannot tell when to call update, so rather i'm unable to tell when all the things are properly initialized.

Here's the code i'm using:

    $('#compare').mergely({
        autoupdate: false,
        width: 'auto',
        height: '600',
        fgcolor: {a:'#ddffdd',c:'#cccccc',d:'#ffdddd'},
        bgcolor: '#fff',
        viewport: true,
        cmsettings: {mode: 'text/plain', readOnly: true, lineWrapping: false, lineNumbers: true},
        lhs: function(setValue) {
                setValue('loading diff...');
                $.ajax(orig1_url, {
                    dataType: 'text',
                    success: setValue
                });

        },
        rhs: function(setValue) {
                setValue('loading diff...');
                $.ajax(orig2_url, {
                    dataType: 'text',
                    success: setValue
                });
        }
    });


Since i load both files async, via ajax. I'm unable to find a proper callback/place so the editor is ready to properly handle `scrollToDiff`

I tried to do it in a setValue wrapper, so once the ajax is finished, but that produces odd results.

**Comment by wickedest** [comment]
Did you look at this link, I sent it in my original response:

https://github.com/wickedest/Mergely/blob/master/editor/editor.js#L193

**Comment by marcinkuzminski** [comment]
Yes i did. It doesn't work with scrollToDiff, i think i really have hard time to find the "...Generally speaking, wait for the lhs/rhs
editors to be initialized, then call scrollToDiff... " Part, since

If i turn of autoupdate, i cannot reliably do either editor.mergely('update'), or editor.mergely('scrollToDiff', 'next'),

I think i'll end up editor the merely code as i did before to make it work.

**Comment by wickedest** [comment]
Looking into this a bit further for you, I realized that it scrolls to the
first change automatically.
https://github.com/wickedest/Mergely/blob/master/lib/mergely.js#L924

If that is not happening for you, then there might be a bug.  I tried it
here with the ajax example, and it "works for me".

**Comment by marcinkuzminski** [comment]
Hi,

So i looked at this code as well, and it's triggered BEFORE i can see content, and diff in mergely, i'm using a 2000 lines file. Maybe that's the reason ?

I can see the _scroll_to_change is called before the diff calculation is made. I'm using autoupdate:true else the diff calculation is breaking in very odd ways (only changes in lhs are visible)

**Comment by wickedest** [comment]
Hi Marcin,

In the code above, you use two separate ajax requests to initialize
Mergely.  The two asynchronous requests can arrive at different times.

Try this; initialize Mergely *without* lhs/rhs:

var editor = $('#compare');
editor.mergely({
    autoupdate: false,
    width: 'auto',
    height: '600',
    fgcolor: {a:'#ddffdd',c:'#cccccc',d:'#ffdddd'},
    bgcolor: '#fff',
    viewport: true,
    cmsettings: {mode: 'text/plain', readOnly: true, lineWrapping:
false, lineNumbers: true}
});

Then wait for the two asynchronous requests.  When done, set the lhs/rhs
editors.

var d1 = $.Deferred();
var d2 = $.Deferred();
$.when(
    $.ajax(orig1_url, {
        dataType: 'text',
        success: function (response) {
            d1.resolve(response);
        }
    }),
    $.ajax(orig2_url, {
        dataType: 'text',
        success: function (response) {
            d2.resolve(response);
        }
    })
).done(function(response1, response2) {
    editor.mergely('lhs', response1);
    editor.mergely('rhs', response2);
});

I wrote the above without any syntax checking.  The scrollToDiff should not
be necessary.

**Comment by marcinkuzminski** [comment]
Tried using deferred, no luck :( But i was expecting this since i had a homebrew code that checked once both lhs and rhs ajax are loaded, and then calling scrollToDiff before and it was still not working.

Could it be that when you set lhs/rhs, the diff calculation is so expensive then scrollToDiff simply is not working ?

**Comment by wickedest** [comment]
No because, if the lhs/rhs get set as above, it will trigger a change, then
synchronously - the change is handled, a diff is calculated, the editors
are scrolled to the first change automatically, then the page is redrawn.

Can you send me, or show me your code?

**Comment by marcinkuzminski** [comment]
Here's the almost complete code:

    var editor = $('#compare');
    editor.mergely({
        autoupdate: true,
        width: 'auto',
        height: '600',
        fgcolor: {a: '#ddffdd', c: '#cccccc', d: '#ffdddd'},
        bgcolor: '#fff',
        viewport: false,
        cmsettings: {
            mode: 'text/plain',
            readOnly: true,
            lineWrapping: false,
            lineNumbers: true
        }
    });

    var lhs = function(deferred) {
                editor.mergely('lhs', 'loading diff...');
                $.ajax(orig1_url, {
                    dataType: 'text',
                    success: function(data) {
                        deferred.resolve(data);
                    }
                });

    };

    var rhs = function(deferred) {
                editor.mergely('rhs', 'loading diff...');
                $.ajax(orig2_url, {
                    dataType: 'text',
                    success: function(data) {
                        deferred.resolve(data);
                    }
                });
    };

    var d1 = $.Deferred();
    var d2 = $.Deferred();
    $.when(
        d1, d2
    ).done(function(lhs_response, rhs_response) {
        editor.mergely('lhs', lhs_response);
        editor.mergely('rhs', rhs_response);
    });
    lhs(d1);
    rhs(d2);

**Comment by wickedest** [comment]
Fixed: https://github.com/wickedest/Mergely/commit/21bd63961b165344888a574fb392fcddf7beec8f#diff-529533c5d49002630583c0b4e7e37860

**Comment by marcinkuzminski** [comment]
Works ! Thanks for your time.
